test/integration/user_test.ml

```diff
@@ -333,6 +333,131 @@ let test_list_moderable_when_there_are_candidates_with_like =
+    )
+;;
+
+let test_update_preference_nothing =
+  integration_test
+  ~about:"update_preferences"
+  ~desc:"If i give empty string to the update form, my infos shouldn't change"
+  (fun _env db ->
+  let open Lwt_util in
+  let*? u =
+    return 
+      (user_for_registration 
+        "gholad"
+        "gholad@gmail.com"
+        "1234567" "1234567")
+  in
+  let*? () = Models.User.register u db in
+  let*? gholad = Models.User.get_by_email "gholad@gmail.com" db in
+  let*? obj = return @@ user_for_update_preferences "" "" in
+  let*? () = Models.User.update_preferences gholad obj db in
+  Models.User.get_by_id gholad.id db
+  )
+(fun user ->
+  let (name, email) =
+    (match user with
+      | Ok gholad -> (gholad.name, gholad.email)
+      | _ -> assert false)
+  in
+  let expected = Ok ("gholad","gholad@gmail.com") in
+  let computed = Ok (name, email) in
+  same (Testable.try_ @@ pair string string) ~expected ~computed
+  )
+;;
+
+
+let test_update_preference_only_username =
+  integration_test
+  ~about:"update_preferences"
+  ~desc:"If i only modify the username, the username should be different and the email shouldn't change"
+  (fun _env db ->
+  let open Lwt_util in
+  let*? u =
+    return 
+      (user_for_registration 
+        "gholad"
+        "gholad@gmail.com"
+        "1234567" "1234567")
+  in
+  let*? () = Models.User.register u db in
+  let*? gholad = Models.User.get_by_email "gholad@gmail.com" db in
+  let*? obj = return @@ user_for_update_preferences "larry_gholad" "" in
+  let*? () = Models.User.update_preferences gholad obj db in
+  Models.User.get_by_id gholad.id db
+  )
+(fun user ->
+  let (name, email) =
+    (match user with
+      | Ok gholad -> (gholad.name, gholad.email)
+      | _ -> assert false)
+  in
+  let expected = Ok ("larry_gholad","gholad@gmail.com") in
+  let computed = Ok (name, email) in
+  same (Testable.try_ @@ pair string string) ~expected ~computed
+  )
+;;
+
+let test_update_preference_both_email_username =
+  integration_test
+  ~about:"update_preferences"
+  ~desc:"If i only the email and username, both should change"
+  (fun _env db ->
+    let open Lwt_util in
+    let*? u =
+      return 
+        (user_for_registration 
+          "gholad"
+          "gholad@gmail.com"
+          "1234567" "1234567")
+    in
+    let*? () = Models.User.register u db in
+    let*? gholad = Models.User.get_by_email "gholad@gmail.com" db in
+    let*? obj = return @@ user_for_update_preferences "larry_gholad" "banane@gmail.com" in
+    let*? () = Models.User.update_preferences gholad obj db in
+    Models.User.get_by_id gholad.id db
+    )
+  (fun user ->
+    let (name, email) =
+      (match user with
+       | Ok gholad -> (gholad.name, gholad.email)
+       | _ -> assert false)
+    in
+    let expected = Ok ("larry_gholad","banane@gmail.com") in
+    let computed = Ok (name, email) in
+    same (Testable.try_ @@ pair string string) ~expected ~computed
+    )
+;;
+
```

Comment:
I suggest to add some test for example, trying to update with an already takes username or email in order to ensure that integrity violation are properly catched.

Reply:
Yes, in fact i just discovered that we can't choose a username which is already taken...., and it changes several things in my code actually, which will be simpler i think. (no need to pass a user to the query update_preferences anymore because the name is pretty much an id, for example)

Reply:
Ok !

Reply:
I started by writing the tests you asked, and the problem is that my solution, which consists of reinserting the user's data if he leaves a form field empty, actually violates the integrity. 

So a better solution would be to call a different SQL query function depending on the non empty field, and require at least one non empty field (actually i can leave both empty)